<a href="https://colab.research.google.com/github/Sasha-Kirsten/Week_3_Churn/blob/main/Week_3_Churn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [648]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import warnings 
warnings.filterwarnings("ignore")

In [649]:
data = 'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv'

In [650]:
!wget $data -O data-week-3.csv

--2022-09-26 20:21:54--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1423529 (1.4M) [text/plain]
Saving to: ‘data-week-3.csv’

data-week-3.csv     100%[===================>]   1.36M  --.-KB/s    in 0.009s  

2022-09-26 20:21:54 (150 MB/s) - ‘data-week-3.csv’ saved [1423529/1423529]



In [651]:
data = pd.read_csv('data-week-3.csv')
data

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  
0            NEAR BAY  
1            NEAR BAY  
2            NEAR BAY  
3            NEAR BAY  
4            NEAR BAY  
...               ...  
20635          INLAND  
20636          INLAND  
20637          INLAND  
20638          INLAND  
20639          INLAND  

[20640 rows x 10 columns]

In [652]:
base = ['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 
        'households', 'median_income', 'median_house_value', 'ocean_proximity']

In [653]:
data =  data.fillna(0)

In [654]:
data.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [655]:
data['rooms_per_household'] = data['total_rooms'].values/data['households'].values
print(data['rooms_per_household'])
# Here create a data values called rooms_per_household, we divide total_rooms values by the households values. 

0        6.984127
1        6.238137
2        8.288136
3        5.817352
4        6.281853
           ...   
20635    5.045455
20636    6.114035
20637    5.205543
20638    5.329513
20639    5.254717
Name: rooms_per_household, Length: 20640, dtype: float64


In [656]:
data['bedrooms_per_room'] = data['total_bedrooms'].values/data['total_rooms'].values
print(data['bedrooms_per_room'])
# Here create a data values called bedrooms_per_room, we divide total_bedrooms values by the total_values values. 

0        0.146591
1        0.155797
2        0.129516
3        0.184458
4        0.172096
           ...   
20635    0.224625
20636    0.215208
20637    0.215173
20638    0.219892
20639    0.221185
Name: bedrooms_per_room, Length: 20640, dtype: float64


In [657]:
data['population_per_household'] = data['population'].values/data['households'].values
print(data['population_per_household'])
# Here create a data values called population_per_household, we divide population values by the households values.

0        2.555556
1        2.109842
2        2.802260
3        2.547945
4        2.181467
           ...   
20635    2.560606
20636    3.122807
20637    2.325635
20638    2.123209
20639    2.616981
Name: population_per_household, Length: 20640, dtype: float64


# Question 1

In [658]:
frequentValue = data['ocean_proximity'].mode
# We are looking for the most category with highest quantity.

In [659]:
print(data['ocean_proximity'].describe())

count         20640
unique            5
top       <1H OCEAN
freq           9136
Name: ocean_proximity, dtype: object


# The answer to question one is <1H Ocean. 

# Question 2

In [660]:
data.corr()

longitude  latitude  housing_median_age  \
longitude                  1.000000 -0.924664           -0.108197   
latitude                  -0.924664  1.000000            0.011173   
housing_median_age        -0.108197  0.011173            1.000000   
total_rooms                0.044568 -0.036100           -0.361262   
total_bedrooms             0.068082 -0.065318           -0.317063   
population                 0.099773 -0.108785           -0.296244   
households                 0.055310 -0.071035           -0.302916   
median_income             -0.015176 -0.079809           -0.119034   
median_house_value        -0.045967 -0.144160            0.105623   
rooms_per_household       -0.027540  0.106389           -0.153277   
bedrooms_per_room          0.084836 -0.104112            0.125396   
population_per_household   0.002476  0.002366            0.013191   

                          total_rooms  total_bedrooms  population  households  \
longitude                    0.044568        0.068082    0.099773    0.055310   
latitude                    -0.036100       -0.065318   -0.108785   -0.071035   
housing_median_age          -0.361262       -0.317063   -0.296244   -0.302916   
total_rooms                  1.000000        0.920196    0.857126    0.918484   
total_bedrooms               0.920196        1.000000    0.866266    0.966507   
population                   0.857126        0.866266    1.000000    0.907222   
households                   0.918484        0.966507    0.907222    1.000000   
median_income                0.198050       -0.007295    0.004834    0.013033   
median_house_value           0.134153        0.049148   -0.024650    0.065843   
rooms_per_household          0.133798        0.002717   -0.072213   -0.080598   
bedrooms_per_room           -0.174583        0.122205    0.031397    0.059818   
population_per_household    -0.024581       -0.028019    0.069863   -0.027309   

                          median_income  median_house_value  \
longitude                     -0.015176           -0.045967   
latitude                      -0.079809           -0.144160   
housing_median_age            -0.119034            0.105623   
total_rooms                    0.198050            0.134153   
total_bedrooms                -0.007295            0.049148   
population                     0.004834           -0.024650   
households                     0.013033            0.065843   
median_income                  1.000000            0.688075   
median_house_value             0.688075            1.000000   
rooms_per_household            0.326895            0.151948   
bedrooms_per_room             -0.573836           -0.238759   
population_per_household       0.018766           -0.023737   

                          rooms_per_household  bedrooms_per_room  \
longitude                           -0.027540           0.084836   
latitude                             0.106389          -0.104112   
housing_median_age                  -0.153277           0.125396   
total_rooms                          0.133798          -0.174583   
total_bedrooms                       0.002717           0.122205   
population                          -0.072213           0.031397   
households                          -0.080598           0.059818   
median_income                        0.326895          -0.573836   
median_house_value                   0.151948          -0.238759   
rooms_per_household                  1.000000          -0.387465   
bedrooms_per_room                   -0.387465           1.000000   
population_per_household            -0.004852           0.003047   

                          population_per_household  
longitude                                 0.002476  
latitude                                  0.002366  
housing_median_age                        0.013191  
total_rooms                              -0.024581  
total_bedrooms                           -0.028019  
population                                0.069863 

total_bedrooms and households:0.965051, This is the right answer

# Make median_house_value binary

In [661]:
mean = data['median_house_value'].mean()

In [662]:
data['above_average'] = np.where(data['median_house_value'] > mean , 1 , 0)

In [663]:
data = data.drop('median_house_value', axis=1)

# Split the data

In [664]:
from sklearn.model_selection import train_test_split 
# We importing train_test_split from sklearn

In [665]:
data_full_train, data_test = train_test_split(data, test_size=0.2,random_state=1)
# We are now spliting the data_test data from original data, only 20%  

In [666]:
data_train, data_validation = train_test_split(data_full_train, test_size=0.25, random_state=1)
# We are now spliting the data into data_train and data_validation,the data_validation is only 25% 
# of the data_full_train 

In [667]:
y_train = data_train.above_average.values
y_validation = data_validation.above_average.values
y_test = data_test.above_average.values
# Here we are creating the y variables, for the trainings set, the validation set and test set.

In [668]:
data_train = data_train.reset_index(drop=True)
data_validation = data_validation.reset_index(drop=True)
data_test = data_test.reset_index(drop=True)
# Here we are reseting the Index of the different data set.

In [669]:
del data_train['above_average']
del data_validation['above_average']
del data_test['above_average']
# Delete the median_house_value variable from the data_train, data_validation, data_test

# Question 3

In [670]:
from sklearn.metrics import mutual_info_score
# importing the mutual info score

In [671]:
categorical = ['longitude', 'latitude', 'housing_median_age', 'total_rooms',
        'total_bedrooms', 'population', 'households', 'median_income', 'ocean_proximity', 'rooms_per_household',
        'bedrooms_per_room', 'population_per_household']

In [672]:
def mutual_info_median_house_value_score(series):
    return mutual_info_score(series, data_full_train.above_average.values)
# making a function for the mutau information. 

In [673]:
mutual_outcome = data_full_train[categorical].apply(mutual_info_median_house_value_score)
mutual_outcome.sort_values(ascending=True)
# looking for the one with the highest rate of mutual information 

housing_median_age          0.013203
households                  0.066152
total_bedrooms              0.071424
ocean_proximity             0.103093
population                  0.136895
latitude                    0.149478
longitude                   0.172332
total_rooms                 0.224974
median_income               0.502303
population_per_household    0.628340
bedrooms_per_room           0.641803
rooms_per_household         0.645401
dtype: float64

# The closest answer to the question 3 is 0.1, from ocean_proximity

# Question 4

In [674]:
from sklearn.feature_extraction import DictVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [675]:
categorical = ['longitude', 'latitude', 'housing_median_age', 'total_rooms',
        'total_bedrooms', 'population', 'households', 'median_income', 'ocean_proximity', 'rooms_per_household',
        'bedrooms_per_room', 'population_per_household']

In [676]:
dicts = data_train[['ocean_proximity']].to_dict(orient='records')

In [677]:
dv = DictVectorizer(sparse=False)

In [678]:
# dv.fit(data_full_train.median_house_value)
dv.fit(dicts)

DictVectorizer(sparse=False)

In [679]:
X_train = dv.transform(dicts)

In [680]:
dv.get_feature_names_out()

array(['ocean_proximity=<1H OCEAN', 'ocean_proximity=INLAND',
       'ocean_proximity=ISLAND', 'ocean_proximity=NEAR BAY',
       'ocean_proximity=NEAR OCEAN'], dtype=object)

In [681]:
list(dv.transform(dicts))

[array([0., 1., 0., 0., 0.]),
 array([0., 1., 0., 0., 0.]),
 array([0., 0., 0., 0., 1.]),
 array([0., 1., 0., 0., 0.]),
 array([0., 1., 0., 0., 0.]),
 array([1., 0., 0., 0., 0.]),
 array([1., 0., 0., 0., 0.]),
 array([1., 0., 0., 0., 0.]),
 array([0., 1., 0., 0., 0.]),
 array([0., 0., 0., 1., 0.]),
 array([0., 0., 0., 0., 1.]),
 array([1., 0., 0., 0., 0.]),
 array([1., 0., 0., 0., 0.]),
 array([0., 1., 0., 0., 0.]),
 array([1., 0., 0., 0., 0.]),
 array([1., 0., 0., 0., 0.]),
 array([1., 0., 0., 0., 0.]),
 array([1., 0., 0., 0., 0.]),
 array([0., 1., 0., 0., 0.]),
 array([1., 0., 0., 0., 0.]),
 array([0., 1., 0., 0., 0.]),
 array([1., 0., 0., 0., 0.]),
 array([1., 0., 0., 0., 0.]),
 array([0., 1., 0., 0., 0.]),
 array([0., 1., 0., 0., 0.]),
 array([0., 0., 0., 1., 0.]),
 array([0., 0., 0., 1., 0.]),
 array([0., 1., 0., 0., 0.]),
 array([1., 0., 0., 0., 0.]),
 array([0., 1., 0., 0., 0.]),
 array([0., 1., 0., 0., 0.]),
 array([0., 1., 0., 0., 0.]),
 array([0., 1., 0., 0., 0.]),
 array([0.

In [682]:
model = LogisticRegression(solver="liblinear", C=1.0, max_iter=1000, random_state=42)
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000, random_state=42, solver='liblinear')

In [683]:
val_dict = data_validation[['ocean_proximity']].to_dict(orient='records')
X_val = dv.transform(val_dict)

y_pred = model.predict(X_val)

In [684]:
accuracy = np.round(accuracy_score(y_validation, y_pred),2)
print(accuracy)

0.66


In [685]:
# The closest answer to the outcome is 0.6

# Question 5

In [686]:
orig_score = accuracy

In [687]:

# subset = categorical.copy()
# subset.remove('ocean_proximity')

for c in categorical:

  subset = categorical.copy()
  subset.remove(c)

  train_dict = data_train[subset].to_dict(orient='records')

  dv = DictVectorizer(sparse=False)
  dv.fit(train_dict)

  X_train = dv.transform(train_dict)

  model = LogisticRegression(solver="liblinear", C=1.0, max_iter=1000, random_state=42)
  model.fit(X_train, y_train)

  val_dict = data_validation[subset].to_dict(orient='records')
  X_val = dv.transform(val_dict)

  y_pred = model.predict(X_val)
  score = accuracy_score(y_validation, y_pred)
  print(c, accuracy - score, score)



longitude -0.15782945736434106 0.8178294573643411
latitude -0.16000968992248055 0.8200096899224806
housing_median_age -0.15855620155038752 0.8185562015503876
total_rooms -0.16388565891472862 0.8238856589147286
total_bedrooms -0.16267441860465115 0.8226744186046512
population -0.153953488372093 0.813953488372093
households -0.1575872093023255 0.8175872093023255
median_income -0.11543604651162787 0.7754360465116279
ocean_proximity -0.14571705426356585 0.8057170542635659
rooms_per_household -0.1629166666666666 0.8229166666666666
bedrooms_per_room -0.1629166666666666 0.8229166666666666
population_per_household -0.16315891472868216 0.8231589147286822


total_bedrooms -0.16267441860465115 0.8226744186046512

total_rooms    -0.16388565891472862 0.8238856589147286 

This is the smallest value, therefore it is the answer: total_rooms.

population     -0.153953488372093 0.813953488372093  

households     -0.1575872093023255 0.8175872093023255

# Question 6

In [688]:
# data['median_house_value'] = np.log1p(data['median_house_value'])

In [689]:
data_full_train = train_test_split(data, test_size=0.2, random_state=1)

In [690]:
data_train, data_validation = train_test_split(data_full_train, test_size=0.33, random_state=11)

In [691]:
y_train = data_train.median_house_value.values
y_validation = data_validation.median_house_value.values
y_test = data_test.median_house_value.values

AttributeError: ignored

In [ ]:
data_train = data_train.reset_index(drop=True)
data_validation = data_validation.reset_index(drop=True)
data_test = data_test.reset_index(drop=True)

In [ ]:
del data_train['median_house_value']
del data_validation['median_house_value']
del data_test['median_house_value']

# Ridge Regression

In [ ]:
train_dict = data_train[categorical].to_dict(orient='records')

In [ ]:
dv = DictVectorizer(sparse=False)
dv.fit(train_dict)

X_train = dv.transform(train_dict)

val_dict = data_validation[categorical].to_dict(orient='records')
X_val = dv.transform(val_dict)


In [ ]:
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error

In [ ]:
for 